bitcoin/src/address/error.rs

```diff
@@ -103,6 +103,12 @@ pub enum ParseError {
     WitnessProgram(witness_program::Error),
     /// Tried to parse an unknown HRP.
     UnknownHrp(UnknownHrpError),
+    /// Legacy address is too long.
+    LegacyAddressTooLong(LegacyAddressTooLongError),
+    /// Invalid base58 payload data length for legacy address.
+    InvalidBase58PayloadLength(InvalidBase58PayloadLengthError),
```

Comment:
Why not treat too long and `InvalidBase58PayloadLenght` the same? They are both invalid lengths just in different layers. It'd be nicer to merge them into a simple `InvalidLength` if we can compute one from the other (which I believe we can) and only present the outer one to the user. (The user is only concerned with number of chars, not decoded bytes.)

Reply:
Warning, hand wavey response. 

Without commenting too much on this specific case, I find complex errors annoying to work with. Error handling is proving to be hard enough as it is using simple types, every time we have more complex errors I end up having to spend ages working out if they conform to all the million nuanced rules we have, this is compounded when errors get used in more places - then when they break the rules it takes much longer to detangle them than if there are just many types. One of the reasons this PR adds multiple errors that are the same "shape" (same fields). I have the gut feeling that forward compatibility will be harder the more we reuse errors in different places also. I think the error code should all be brain dead simple and ridiculously uniform if we are to have any hope of limiting mistakes. Its a moving target still, unfortunately, I can do error stuff mechanically in the afternoon if its brain dead easy - if its complex it means I have to do it earlier in the day when I have more brain power and it also needs more thought when reviewing. Long wall of text, sorry, I've spent a lot of time writing error stuff lately :)

Reply:
I tend to agree -- and to the extent possible, as a user I like when every error is only constructible from one place. Then I can trace backward from the rust-bitcoin source to figure out where the error actually came from (after grepping to reverse the error message into an error variant).

While conceptually "invalid length" always means that the string was somehow an invalid length, it's useful to know which layer of the decoding complained.

Reply:
@tcharding is it possible that you're confusing this with an objection against same-shaped errors? Because I'm not objecting to that. Same-shaped errors with different semantics are fine. Differently-shaped errors with the same semantics are not really that great.

I know it can be annoying but really, programming is hard anyway. If we merge the errors in thee library we can save downstream users from verbose matching. I tend to prefer making libraries internally complex if it can make downstream crates less complex because there are multiple downstream projects.

@apoelstra by "users" here I mean the people who are not programmers. I see no reason why they should concern themselves with layers. If you need to debug things we could make the debug representation print out the source or we could even add optional backtrace.

Reply:
>@apoelstra by "users" here I mean the people who are not programmers

You've defined "user" in a way that can't possibly include me, which isn't fair. I also use software. In basically every software I use when I encounter an error of some sort I dig into the source code.

If there is no source code (e.g. the opaque "Contact Key Verification" error that you get when you try to enable the new iMessage fingerprint verification tool) then generally I behave like a non-technical person by giving up and doing something else that doesn't involve computers.

> If you need to debug things we could make the debug representation print out the source or we could even add optional backtrace.

I really like this idea. We could compile out that stuff when `debug_assertions` is off.

The backtrace API isn't stable until 1.65 so something to revisit, not related to this PR or anything soon.

BTW this is an example of something we could add where @tcharding's constructors would be unaffected :). They would just need to have `#[track_caller]` on and they could conditionally internally generate backtraces.

Reply:
> You've defined "user" in a way that can't possibly include me, which isn't fair. I also use software. In basically every software I use when I encounter an error of some sort I dig into the source code.

Fair, I do this too. :D

Maybe the error messages can be slightly different such that it doesn't bother users but allows debugging.

> We could compile out that stuff when `debug_assertions` is off.

I'd rather use a different feature flag if we want to provide a flag at all. The `Backtrace` type already handles not capturing when `RUST_BACKTRACE` is not set

> The backtrace API isn't stable until 1.65 so something to revisit

We can use `cfg` and/or replace it with `core::panic::Location` in the meantime (and maybe we should do it anyway because panic location is `no_std` while backtrace isn't).

There's also external library for this but we can't make it conditional on Rust version so I'd rather not use it.

Reply:
Is this thread resolved @Kixunil or do you still want to see changes to the introduced error types?